File: python/random_walk.py
import numpy as np
import time


class Point:
    def __init__(self, x, y, z):
        self.x = x
        self.y = y
        self.z = z

def random_walk(num_steps, step_size):
    rng = np.random.default_rng()
    current = Point(0.0, 0.0, 0.0)

    for _ in range(num_steps):
        theta = rng.uniform(0.0, 2.0 * np.pi)
        phi = rng.uniform(0.0, np.pi)
        dx = step_size * np.cos(theta) * np.sin(phi)
        dy = step_size * np.sin(theta) * np.sin(phi)
        dz = step_size * np.cos(phi)

        current.x += dx
        current.y += dy
        current.z += dz

    return current

def mean_squared_displacement(k, n, l):
    total_msd = 0.0

    for _ in range(k):
        final_point = random_walk(n, l)
        dx = final_point.x
        dy = final_point.y
        dz = final_point.z

        msd = dx**2 + dy**2 + dz**2
        total_msd += msd

    return total_msd / k

def main():
    k_values = [10, 50]
    n_values = [10000000]
    l = 1.0
    for n in n_values:
        for k in k_values:
            start_time = time.perf_counter()

            msd = mean_squared_displacement(k, n, l)
            
            end_time = time.perf_counter()

            print(f"Mean Square Displacement for k = {k} replicates, num_steps = {n}: {msd}, time_taken = {end_time-start_time}")

if __name__ == "__main__":
    main()

File: golang/random_walk.go
package main

import (
	"fmt"
	"math"
	"math/rand"
	"time"
)

type Point struct {
	x, y, z float64
}

func randomWalk(numSteps int, stepSize float64) Point {
	rng := rand.New(rand.NewSource(time.Now().UnixNano()))
	current := Point{0.0, 0.0, 0.0}

	for i := 0; i < numSteps; i++ {
		theta := rng.Float64() * 2.0 * math.Pi
		phi := rng.Float64() * math.Pi
		dx := stepSize * math.Cos(theta) * math.Sin(phi)
		dy := stepSize * math.Sin(theta) * math.Sin(phi)
		dz := stepSize * math.Cos(phi)

		current.x += dx
		current.y += dy
		current.z += dz
	}

	return current
}

func meanSquaredDisplacement(k, n int, l float64) float64 {
	totalMSD := 0.0

	for i := 0; i < k; i++ {
		finalPoint := randomWalk(n, l)
		dx := finalPoint.x
		dy := finalPoint.y
		dz := finalPoint.z

		msd := dx*dx + dy*dy + dz*dz
		totalMSD += msd
	}

	return totalMSD / float64(k)
}

func main() {
	kValues := []int{10,100, 1000}
	nValues := []int{10000, 20000, 50000}
    l := 1.0

	for _, n := range nValues {
		for _, k := range kValues {
			startTime := time.Now()

			msd := meanSquaredDisplacement(k, n, l)

			endTime := time.Now()

			fmt.Printf("Mean Square Displacement for k = %d replicates, num_steps = %d: %f, time_taken = %v\n", k, n, msd, endTime.Sub(startTime))
		}
	}
}

